slack-api.ts

```diff
@@ -0,0 +1,157 @@
+ */
+
+import { HttpService } from '@nestjs/axios';
+import { Logger } from '@nestjs/common';
+import axios from 'axios';
+import { firstValueFrom } from 'rxjs';
+
+import { Slack } from './types';
+import { verifySlackRequest } from './verify-request';
+
+export class SlackApi {
+  private httpService: HttpService;
+
+  //TODO: handle api errors
+  constructor(
+    access_token: string,
+    private signing_secret: string,
+  ) {
+    this.buildHttpService(access_token);
+  }
+
+  setAccessToken(access_token: string) {
+    Logger.verbose('Access token updated', 'Slack Api');
+    this.buildHttpService(access_token);
+  }
+
+  setSigningSecret(signing_secret: string) {
+    Logger.verbose('Signing secret updated', 'Slack Api');
+    this.signing_secret = signing_secret;
+  }
+
+  private buildHttpService(access_token: string) {
+    if (!access_token) {
+      Logger.error('Access token is missing', 'Error: Slack Api');
+    }
+    const axiosInstance = axios.create({
+      baseURL: 'https://slack.com/api',
+      headers: {
+        'Content-Type': 'application/json; charset=UTF-8',
+        Authorization: `Bearer ${access_token}`,
+      },
+    });
+    axiosInstance.interceptors.response.use(
+      (response) => {
+        if (response.data.ok === false) {
+          debugger; //
+          Logger.error(
+            `${response.data.error}: ${response.data.errors?.join(', ')}`,
+            'Error: Slack API',
+          );
+        }
+        return response;
+      },
+      (error) => {
+        debugger; //
+        Logger.error(error, 'Error: Slack Api');
+      },
+    );
+    this.httpService = new HttpService(axiosInstance);
+  }
+
+  public verifySignature(req: any) {
+    try {
+      verifySlackRequest({
+        signingSecret: this.signing_secret,
+        //body: req.body,// take the raw body instead
+        body: req.rawBody,
+        headers: {
+          'x-slack-signature': req.headers['x-slack-signature'],
+          'x-slack-request-timestamp': req.headers['x-slack-request-timestamp'],
+        },
+      });
+      return true;
+    } catch (e) {
+      Logger.error(e, 'Error: Slack Api');
+    }
+  }
+
+  async getUserInfo(userForeingId: string): Promise<Slack.User> {
+    return (
+      (
+        await firstValueFrom(
+          this.httpService.get(Slack.ApiEndpoint.usersInfo, {
+            params: { user: userForeingId },
+          }),
+        )
+      ).data as Slack.UsersInfoResponse
+    ).user;
+  }
+
+  async sendMessage(message: any, channel) {
+    return await firstValueFrom(
+      this.httpService.post(Slack.ApiEndpoint.chatPostMessage, {
+        channel,
+        ...message,
+      }),
+    );
+  }
+
+  async getUploadURL(
+    fileName: string,
+    size: number,
+  ): Promise<Slack.UploadUrlData> {
+    return (
+      await firstValueFrom(
+        this.httpService.get(Slack.ApiEndpoint.getUploadURL, {
+          params: { filename: fileName, length: size },
+        }),
+      )
+    ).data;
```

Comment:
```suggestion
    const { data } =  
      await this.httpService.axiosRef.get(Slack.ApiEndpoint.getUploadURL, {
          params: { filename: fileName, length: size },
        });
        
      return data;
```

Can you do the same for all the other calls ?